### PR TITLE
Fast histogram observer

### PR DESCRIPTION
**Pull request by raghuramank100**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29790 Fast histogram observer**
Summary:

 10x speed up of histogram observers by rewriting histogram combination routine.
 Previously this was done as explicit bilinear interpolation.
 Now this is done as a sample rate conversion operation, where we achieve
 resampling by an upsampling (zero-order hold) followed by box filtering and
 downsampling.

Test Plan:
import torch
import time
import numpy as np
from torch.quantization.observer import HistogramObserver

X = torch.randn(1,1,224,224)

obs = HistogramObserver(2048)
acc_time = 0
for i in range(100):
   X = torch.randn(10,1,320,320)
   start = time.time()
   obs(X)
   #obs.forward_new(X)
   acc_time = acc_time + time.time()-start
print(acc_time)

Before change:
6.9

After change:
0.6

Resnet-18 accuracy is unchanged with the faster histogram observer.
Acc1 = 69.4

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18508562](https://our.internmc.facebook.com/intern/diff/D18508562)

### CODE REVIEW

**Review by lly-zero-one — COMMENTED** [review]
NVM, it seems you already added it.

File: torch/quantization/observer.py [hunk 1]
```diff
@@ -567,6 +567,7 @@ def __init__(self, bins=2048, dtype=torch.quint8,
         self.min_val = None
         self.max_val = None
         self.dst_nbins = 2 ** torch.iinfo(self.dtype).bits
+        self.upsample_rate = 128
```
Comment (hx89) [inline torch/quantization/observer.py]: May add comment on why choose 128 here? And could make upsample_rate as an input argument.

**Review by hx89 — COMMENTED** [review]
This is great idea and huge speedup!

**Review by hx89 — APPROVED** [review]
LGTM!
